service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java

```diff
@@ -189,8 +189,15 @@ public Integer getNumberOfSamplesInMolecularProfile(String molecularProfileId) {
             samples = sampleService.fetchSamples(studyIds, sampleIds, "ID");
         }
 
-        List<GeneMolecularAlteration> molecularAlterations = molecularDataRepository
-            .getGeneMolecularAlterationsInMultipleMolecularProfiles(distinctMolecularProfileIds, entrezGeneIds, projection);
+        List<GeneMolecularAlteration> molecularAlterations = new ArrayList<>();
+        // query each entrezGeneId separately so they can be cached
+        for (Integer entrezGeneId: entrezGeneIds) {
+            List<Integer> singleGene = Arrays.asList(new Integer[]{ entrezGeneId });
+            molecularAlterations.addAll(
+                molecularDataRepository
+                    .getGeneMolecularAlterationsInMultipleMolecularProfiles(distinctMolecularProfileIds, singleGene, projection)
+            );
+        }
```

Comment:
This is fundementally fine, but there's a few small things we can improve:
`Arrays.asList(new Integer[]{ entrezGeneId });`-> `Arrays.asList(entrezGeneId);` Java has varargs. You are right that an array of the same type is equivalent, but this looks a bit cleaner.

`Arrays.asList` vs `Collections.singletonList`. Use the former for lists with multiple elements. Use the latter for singleton lists. 

Iterating using a for loop vs streaming. Java streams are pretty nice. They're lazily evaluated, so you can split things into easy to read sections without taking a performance hit. Generally, if I can solve something in a stream without any side effects, I do. This is just a flatmap and a collect.

Putting all those things together, here's what I would recommend:
```suggestion
        // query each entrezGeneId separately so they can be cached
        List<GeneMolecularAlteration> molecularAlterations = entrezGeneIds.stream()
            .flatMap(gene -> molecularDataRepository.getGeneMolecularAlterationsInMultipleMolecularProfiles(
                    distinctMolecularProfileIds, Collections.singletonList(gene), projection
                ).stream()
            )
            .collect(Collectors.toList());
```